### PR TITLE
Add stereo, sample rate, loss balancer support

### PR DESCRIPTION
**Pull request by leoauri**
Hi there, I have worked with the repo for training and extended some features:
- 48kHz
- stereo
- use mp3s for dataset
- use loss balancer
- example Slurm sbatch files
- small fixes

Feedback welcome! Thanks for your work :)

### CODE REVIEW

**Comment by ZhikangNiu** [comment]
@leoauri it's absolutely so cool and I will review your pull request!

**Comment by ZhikangNiu** [comment]
@leoauri Can you share some reconstruction sample in this pr or some public checkpoint if available. I just want to check this pr can reconstruct audio samples normally.

**Comment by leoauri** [comment]
Hi @ZhikangNiu, I'm mid-training now. The losses generally look like they are dropping, so I think using two `.backwards()` calls is working. Maybe the learning rate default 3e-4 is too high. Reconstructions sound still noisy but the model seems to be training. I'm uploading checkpoint and reconstruction samples rn. Thanks!

**Comment by leoauri** [comment]
[Here](http://content.leoauri.com/stuff/encodec_training_demo/) are model checkpoints & reconstruction samples. I loaded it with:
```
model = EncodecModel._get_model(
            [1.5, 3., 6., 12., 24.], 48_000, 2,
            causal=False, model_norm='time_group_norm', 
            audio_normalize=True,
            segment=1., name='encodec_48khz_reproduce',
            )
state_dict = torch.load(modeldir/'bs8_cut96000_length0_epoch67_lr0.0003.pt', map_location=torch.device('cpu'))['model_state_dict']
model.load_state_dict(state_dict)

model.set_target_bandwidth(24.0)
```

**Comment by ZhikangNiu** [comment]
> Hi @ZhikangNiu, I'm mid-training now. The losses generally look like they are dropping, so I think using two `.backwards()` calls is working. Maybe the learning rate default 3e-4 is too high. Reconstructions sound still noisy but the model seems to be training. I'm uploading checkpoint and reconstruction samples rn. Thanks!

Yeah, it sounds not very perfect. I think you can try decrease tensor_cut so that you can increase your batch size (for example 1s, if your audio is 48khz, it means that the tensor cut should be 48000). I think when traing gan, it needs bigger batch size.

File: audio_to_mel.py
Comment (ZhikangNiu) [inline audio_to_mel.py:37]: process two channel audio？

**Comment by leoauri** [comment]
> > Hi @ZhikangNiu, I'm mid-training now. The losses generally look like they are dropping, so I think using two `.backwards()` calls is working. Maybe the learning rate default 3e-4 is too high. Reconstructions sound still noisy but the model seems to be training. I'm uploading checkpoint and reconstruction samples rn. Thanks!
> 
> Yeah, it sounds not very perfect. I think you can try decrease tensor_cut so that you can increase your batch size (for example 1s, if your audio is 48khz, it means that the tensor cut should be 48000). I think when traing gan, it needs bigger batch size.

I'm training now with larger batch size, let's see how it goes.

I added commits to fix double logging, save current reconstruction examples, remove corrupt files from dataset, update discriminator with given probability (I found discriminator was overpowering generator at higher sample rate so implemented as they describe in paper).

File: train_multi_gpu.py
Comment (ZhikangNiu) [inline train_multi_gpu.py:332]: When test model on epoch 0, I think it couldn't reconstruction waveform. So shell we add `if epoch == 0` and don't save audio samples. I think epoch 0 test function is just to check model can run successfully？ @leoauri  BTW, i think we should save reconstruction sample and GT sample under the `sp / {epoch}-samples` folder

**Comment by ZhikangNiu** [comment]
I think there are too many shell scripts and slurm scripts in this dictionary. Could you make a new folder `scripts` and move them in this folder. I think you need to fix some minor bug ( or you can make a new folder `slurm_scripts` and move `*.sbatch` script in this folder @leoauri

**Comment by leoauri** [comment]
I'll leave it up to you to organise the whole repo as you see fit. Just curious, did you have a reason to move the main encodec library to root?

Then you mentioned some minor bug, you want to elaborate?

**Comment by ZhikangNiu** [comment]
> I'll leave it up to you to organise the whole repo as you see fit. Just curious, did you have a reason to move the main encodec library to root?
> 
> Then you mentioned some minor bug, you want to elaborate?

Ok, I will re-organise the repo after merging your pr. Minor bug i think when we re-organise the repo, maybe there are some path problem.

**Comment by leoauri** [comment]
Right, I threw all the scripts into a folder. The slurm works still by running `sbatch scripts/train.sbatch` from repo root. Didn't test the shell scripts.

**Comment by ZhikangNiu** [comment]
> Right, I threw all the scripts into a folder. The slurm works still by running `sbatch scripts/train.sbatch` from repo root. Didn't test the shell scripts.

thx leo! Can you share some public 48khz model checkpoint or example if available (It's unnecessary.

**Comment by leoauri** [comment]
> > Right, I threw all the scripts into a folder. The slurm works still by running `sbatch scripts/train.sbatch` from repo root. Didn't test the shell scripts.
> 
> thx leo! Can you share some public 48khz model checkpoint or example if available (It's unnecessary.

Yes, I'm mid-training, but sounding pretty good: https://ufile.io/f/az2fu

This is 3 days on one A100.

**Comment by ZhikangNiu** [comment]
> > > Right, I threw all the scripts into a folder. The slurm works still by running `sbatch scripts/train.sbatch` from repo root. Didn't test the shell scripts.
> > 
> > 
> > thx leo! Can you share some public 48khz model checkpoint or example if available (It's unnecessary.
> 
> Yes, I'm mid-training, but sounding pretty good: https://ufile.io/f/az2fu
> 
> This is 3 days on one A100.

yeah sounds not bad

**Review by ZhikangNiu — APPROVED** [review]

**Comment by ZhikangNiu** [comment]
Yeah leo, I have merged your pr in our repo. So can you update README with you pr updates?

**Pull request merged**
